circuit/src/builder/circuit_builder.rs

```diff
@@ -79,26 +84,67 @@ where
     }
 
     /// Enables a non-primitive operation type on this builder.
-    pub fn enable_op(&mut self, op: NonPrimitiveOpType, cfg: crate::op::NonPrimitiveOpConfig) {
+    pub fn enable_op(&mut self, op: NonPrimitiveOpType, cfg: crate::op::NonPrimitiveOpConfig<F>) {
         self.config.enable_op(op, cfg);
     }
 
     /// Enables Poseidon permutation operations (one perm per table row).
     ///
-    /// The current implementation only supports extension degree D=4.
-    pub fn enable_poseidon_perm<Config: Poseidon2Params>(
-        &mut self,
-        trace_generator: TraceGeneratorFn<F>,
-    ) where
-        F: CircuitField,
+    /// The current implementation only supports extension degree D=4 and WIDTH=16.
+    ///
+    /// # Arguments
+    /// * `trace_generator` - Function to generate Poseidon trace from circuit and witness
+    /// * `perm` - The Poseidon permutation to use for execution
+    pub fn enable_poseidon_perm<Config, P>(&mut self, trace_generator: TraceGeneratorFn<F>, perm: P)
+    where
+        Config: Poseidon2Params,
+        F: CircuitField + ExtensionField<Config::BaseField>,
+        P: Permutation<[Config::BaseField; 16]> + Clone + Send + Sync + 'static,
     {
-        // Hard gate on D=4 to avoid silently accepting incompatible configs.
+        // Hard gate on D=4 and WIDTH=16 to avoid silently accepting incompatible configs.
         assert!(
             Config::D == 4,
             "Poseidon perm op only supports extension degree D=4"
         );
+        assert!(
+            Config::WIDTH == 16,
+            "Poseidon perm op only supports WIDTH=16"
+        );
+
+        // Build exec closure that:
```

Comment:
You can also make `PoseidonPermExecutor` generic over the permutation adding also a field permutation, and do all this directly on `exec`.